autoload/codefmt/rubocop.vim

```diff
@@ -0,0 +1,70 @@
+" Copyright 2023 Google Inc. All rights reserved.
+"
+" Licensed under the Apache License, Version 2.0 (the "License");
+" you may not use this file except in compliance with the License.
+" You may obtain a copy of the License at
+"
+"     http://www.apache.org/licenses/LICENSE-2.0
+"
+" Unless required by applicable law or agreed to in writing, software
+" distributed under the License is distributed on an "AS IS" BASIS,
+" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+" See the License for the specific language governing permissions and
+" limitations under the License.
+
+let s:plugin = maktaba#plugin#Get('codefmt')
+
+
+""
+" @private
+" Formatter: rubocop
+function! codefmt#rubocop#GetFormatter() abort
+  let l:formatter = {
+      \ 'name': 'rubocop',
+      \ 'setup_instructions': 'Install rubocop ' .
+          \ '(https://rubygems.org/gems/rubocop).'}
+
+  function l:formatter.IsAvailable() abort
+    return executable(s:plugin.Flag('rubocop_executable'))
+  endfunction
+
+  function l:formatter.AppliesToBuffer() abort
+    return &filetype is# 'eruby' || &filetype is# 'ruby'
+  endfunction
+
+  ""
+  " Reformat the current buffer with rubocop or the binary named in
+  " @flag(rubocop_executable), only targeting the range between {startline} and
+  " {endline}.
+  " @throws ShellError
+  function l:formatter.FormatRange(startline, endline) abort
+    " See flag explanations at:
+    " https://docs.rubocop.org/rubocop/1.51/usage/basic_usage.html
+    let l:cmd = [s:plugin.Flag('rubocop_executable'), '--stdin', @%, '-a', '--no-color', '-fq', '-o', '/dev/null']
+
+    call maktaba#ensure#IsNumber(a:startline)
```

Comment:
Where this is basically a copy-paste of parts of `codefmt#formatterhelpers#AttemptFakeRangeFormatting`, do you think for maintainability's sake it would be possible to factor this out as a named helper like `s:AttemptFakeRangeFormattingIgnoringFirstLine` to at least try to keep the common logic as one-to-one with AttemptFakeRangeFormatting as possible (and eventually consolidate those somehow)?

Reply:
I stared at this for a few minutes, and I don't think we should pull this out as a separate function, because the error handling in this case is unusual. We can't just use Call() in the function and catch the exception, because that throws away the result. But I'd expect most callers to want normal error handling.

I could generalize AttemptFakeRangeFormatting with two new parameters, e.g. ignorefirstnlines and ignorefailures, if you think it's worth the complexity for other callers. Then you end up with something like:


function! codefmt#formatterhelpers#AttemptFakeRangeFormatting(
...
  let l:result = ignorefailures ? maktaba#syscall#Create(a:cmd).WithStdin(l:input).Call(0) : maktaba#syscall#Create(a:cmd).WithStdin(l:input).Call()
  let l:formatted = split(l:result.stdout, "\n")
  " Special case empty slice: neither l:lines[:0] nor l:lines[:-1] is right.
  let l:before = a:startline > 1 ? l:lines[ : a:startline - 2] : []
  let l:full_formatted = l:before + l:formatted[ignorefirstnlines :] + l:lines[a:endline :]
...

Reply:
... I went ahead and did this; let me know which version you prefer.